signature-specification.md

```diff
@@ -0,0 +1,56 @@
+# Signature Specification
+
+This document specifies the signature format and its protocol.
+
+
+## Terminology
+
+`SignatureEnvelope` (![\varepsilon](https://render.githubusercontent.com/render/math?math=%5Ctextstyle+%5Cvarepsilon)): The standard data structure for storing a signed message. Signature envelope consist of following components:
+
+- Payload (![m](https://render.githubusercontent.com/render/math?math=%5Ctextstyle+m)): The data that is integrity protected - e.g. container image descriptor.
+- Signed attributes (![\mu](https://render.githubusercontent.com/render/math?math=%5Ctextstyle+%5Cmu)): The signature metadata that is integrity protected - e.g. signature expiration time, signing time, etc.
+- Unsigned attributes (![\hat\mu](https://render.githubusercontent.com/render/math?math=%5Ctextstyle+%5Chat%5Cmu).): The signature metadata that is not integrity protected - e.g. keyid, publickey, etc.
+- Cryptographic signatures (![s](https://render.githubusercontent.com/render/math?math=%5Ctextstyle+s)): The digital signatures computed on payload and signed attributes.
+
+A signature ![s = \mathbf{Sign}(K_\text{priv}, m, \mu)](https://render.githubusercontent.com/render/math?math=%5Ctextstyle+s+%3D+%5Cmathbf%7BSign%7D%28K_%5Ctext%7Bpriv%7D%2C+m%2C+%5Cmu%29) where ![K_\text{priv}](https://render.githubusercontent.com/render/math?math=%5Ctextstyle+K_%5Ctext%7Bpriv%7D) stands for signing key.
+A signature envelope ![\varepsilon_n = \{ m, (\mu_0, \hat\mu_0, s_0), \dots, (\mu_{n-1}, \hat\mu_{n-1}, s_{n-1}) \}](https://render.githubusercontent.com/render/math?math=%5Ctextstyle+%5Cvarepsilon_n+%3D+%5C%7B+m%2C+%28%5Cmu_0%2C+%5Chat%5Cmu_0%2C+s_0%29%2C+%5Cdots%2C+%28%5Cmu_%7Bn-1%7D%2C+%5Chat%5Cmu_%7Bn-1%7D%2C+s_%7Bn-1%7D%29+%5C%7D) is a compact form of `n` signatures signing the same payload `m`.
+
+![A complete signature in OCI registry](media/signature-structure.svg)
+
+
+## Requirements
+
+### Signature Envelope and Signature
+
+1. The signature envelope payload MUST contain a [descriptor](https://github.com/opencontainers/image-spec/blob/master/descriptor.md#properties) of the target manifest.
+    1. A descriptor MUST contain `mediaType`, `digest`, `size` fields.
+    2. A descriptor MAY contain the `artifactType` field for artifact manifests, or the `config.mediaType` for `oci.image` based manifests.
+    3. descriptor MAY contain a `annotations` field.
+2. The signature envelope MUST be self contained to validate the integrity of the payload and signed attributes.

```

Comment:
I have explored the idea of X.509 certificate blob caching in an unrelated project of mine (https://github.com/Devolutions/picky-rs/blob/master/docs/PICKY.adoc). It would be relatively simple to push individual X.509 certificates in ASN.1 DER binary format as raw blobs, and then refer to the certificate chain as a simple array of digests. In my implementation I have used multibase-encoded multihashes, but the regular digest notation used in OCI manifests should probably prevail. In all cases, please use the binary representation of the certificate, it's just simpler for blobs - all textual representations only exist to transfer the binary format over text. This being said, my own PKI server would use the certificate digest as the key and support calls to fetch the certificate in various formats - raw binary or PEM, to accomodate clients.